lightningd/peer_control.c

```diff
@@ -1107,6 +1107,47 @@ peer_connected_hook_deserialize(struct peer_connected_hook_payload *payload,
 	return true;
 }
 
+/* Compare and store `remote_addr` and the `peer_id` that reported it.
+ * If new address was reported by at least one other, do node_announcement */
+static void update_remote_addr(struct lightningd *ld,
+			       struct wireaddr *remote_addr,
+			       struct node_id peer_id)
+{
+	/* init pointers first time */
+	if (ld->remote_addr_v4 == NULL) {
+		ld->remote_addr_v4 = tal(ld, struct wireaddr);
+		*ld->remote_addr_v4 = *remote_addr;
+		ld->remote_addr_v4_peer = peer_id;
+	}
+	if (ld->remote_addr_v6 == NULL) {
+		ld->remote_addr_v6 = tal(ld, struct wireaddr);
+		*ld->remote_addr_v6 = *remote_addr;
+		ld->remote_addr_v6_peer = peer_id;
+	}
+
+	switch (remote_addr->type) {
+	case ADDR_TYPE_IPV4:
+		if (node_id_eq(&ld->remote_addr_v4_peer, &peer_id))
+			break;
+		*ld->remote_addr_v4 = *remote_addr;
+		ld->remote_addr_v4_peer = peer_id;
+		break;
+	case ADDR_TYPE_IPV6:
+		if (node_id_eq(&ld->remote_addr_v6_peer, &peer_id))
+			break;
+		*ld->remote_addr_v6 = *remote_addr;
+		ld->remote_addr_v6_peer = peer_id;
+		break;
```

Comment:
So, if the same node gives you a new address, you ignore it?  That seems weird.

I'd expect this code to be:

```
        /* Update any v4/v6 address if they've given one */
        switch (remote_addr->type) {
	case ADDR_TYPE_IPV4:
                tal_free(ld->remote_addr_v4);
                ld->remote_addr_v4 = tal_dup(ld, struct wireaddr, remote_addr);
                ld->remote_addr_v4_peer = *peer_id;
       ...
```

Reply:
Yes, I need to store the latest address of the same peer.
I rewrote the block.